lib/teleagent/lazy.go

```diff
@@ -0,0 +1,112 @@
+package teleagent
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gravitational/trace"
+
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+)
+
+// lazyAgent is an agent.Agent instance which is lazily initialized
+// via a call to an AgentGetter.  Helpful for dealing with APIs
+// which require, but may not actually use, an agent.
+type lazyAgent struct {
+	ctx  context.Context
+	get  AgentGetter
+	once sync.Once
+	ref  agent.Agent
+	err  error
+}
+
+// Lazy wraps an AgentGetter in an agent.Agent interface which lazily
+// initializes the agent on first method invocation.
+func Lazy(ctx context.Context, get AgentGetter) agent.Agent {
```

Comment:
https://github.com/golang/go/wiki/CodeReviewComments#interfaces return a specific exported type `*LazyAgent` (or `*Lazy` and rename this func to `NewLazy`)